File: competitive-programming/src/edu/algos/li17_graph/_ShortestPath_L1631_PathWithMinimumEffort.java
package edu.algos.li17_graph;

import java.util.Arrays;
import java.util.PriorityQueue;

public class _ShortestPath_L1631_PathWithMinimumEffort {
    public static void main(String[] args) {
        _ShortestPath_L1631_PathWithMinimumEffort obj = new _ShortestPath_L1631_PathWithMinimumEffort();

        int[][] heights = new int[][]{{1,2,2},{3,8,2},{5,3,5}};
        System.out.println(obj.minimumEffortPath(heights));
    }

    // Problem: https://leetcode.com/problems/path-with-minimum-effort/
    // Idea: Apply Dijkstra with dist as abs effort diff.
    public int minimumEffortPath(int[][] heights) {
        int m=heights.length, n=heights[0].length;

        PriorityQueue<int[]> pq = new PriorityQueue<>((e1, e2) -> e1[0]-e2[0]);
        int[][] dist = new int[m][n];
        for(int i=0; i<m; i++) 
            Arrays.fill(dist[i], Integer.MAX_VALUE);

        dist[0][0] = 0;
        pq.add(new int[]{0, 0, 0}); // {dist, row, col}
        while(!pq.isEmpty()){
            int[] top = pq.remove();
            int effort = top[0], row = top[1], col = top[2];
            if(row == m-1 && col == n-1)
                return effort;

            int[][] directions = new int[][]{{0,-1}, {-1, 0}, {0, 1}, {1, 0}};
            for(int direction[]: directions){
                int nrow = row+direction[0], ncol = col+direction[1];
                if(nrow>=0 && nrow<m && ncol>=0 && ncol<n){
                    int newEffort = Math.max(effort, Math.abs(heights[nrow][ncol] - heights[row][col]));
                    if(newEffort < dist[nrow][ncol]){
                        dist[nrow][ncol] = newEffort;
                        pq.add(new int[]{newEffort, nrow,ncol});
                    }
                }
            }
        }

        return 0;
    }
}

File: dsa/patterns/13_graphs/python/shortest_path_l1631_path_with_minimum_effort.py
import heapq
import math


"""
Problem: https://leetcode.com/problems/path-with-minimum-effort/
Idea:
Apply Dijkstra's algorithm where the distance is defined as the
maximum absolute height difference encountered along the path.

Instead of summing weights, we minimize the maximum edge cost on the path.

Time Complexity: O(m * n * log(m * n))
Space Complexity: O(m * n)
"""
def minimum_effort_path(heights):
    m, n = len(heights), len(heights[0])

    dist = [[math.inf] * n for _ in range(m)]
    dist[0][0] = 0

    pq = []
    heapq.heappush(pq, (0, 0, 0))   # (effort, row, col)

    while pq:
        effort, row, col = heapq.heappop(pq)

        if row == m - 1 and col == n - 1:
            return effort

        for dr, dc in ((0, -1), (-1, 0), (0, 1), (1, 0)):
            nrow, ncol = row + dr, col + dc
            if is_safe(m, n, nrow, ncol):
                next_effort = max(
                    effort,
                    abs(heights[nrow][ncol] - heights[row][col])
                )

                if next_effort < dist[nrow][ncol]:
                    dist[nrow][ncol] = next_effort
                    heapq.heappush(pq, (next_effort, nrow, ncol))

    return 0


def is_safe(m, n, r, c):
    return 0 <= r < m and 0 <= c < n


# one test (same as Java main)
if __name__ == "__main__":
    heights = [
        [1, 2, 2],
        [3, 8, 2],
        [5, 3, 5]
    ]
    print(minimum_effort_path(heights))  # 2
